### PR TITLE
[pkg/ottl] Changed replacement string to be a path expression to a string telemetry field or a literal string

### PR DESCRIPTION
**Pull request by rnishtala-sumo**
**Description:** 
- Changed replacement_pattern (replacement) string to be a path expression to a string telemetry field or a literal string
- Added new hash functions to the factory map

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22787
**Testing:** 
- Tested locally using the following processors
```
  attributes/example:
    actions:
      - key: message
        pattern: "^device=(?P<device_name>\\w+)$"
        action: extract
  transform/replace:
    log_statements:
      - context: log
        statements:
          - set(attributes["device_name"], SHA256(attributes["device_name"]))
          - replace_pattern(attributes["message"], "^device=[0-9A-Za-z]+", attributes["device_name"])
```
- Updated unit tests

### CODE REVIEW

File: processor/transformprocessor/internal/common/functions.go [hunk 1]
```diff
@@ -33,6 +33,9 @@ func Functions[K any]() map[string]ottl.Factory[K] {
 		ottlfuncs.NewDeleteMatchingKeysFactory[K](),
 		ottlfuncs.NewMergeMapsFactory[K](),
 		ottlfuncs.NewLogFactory[K](),
+		ottlfuncs.NewFnvFactory[K](),
+		ottlfuncs.NewSHA1Factory[K](),
+		ottlfuncs.NewSHA256Factory[K](),
```
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/common/functions.go:38]: Please remove this and add it to its own PR or leave a review on https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23190

**Review by TylerHelmuth — COMMENTED** [review]
Can you update all the replace_* functions to accept a StringGetter

File: processor/transformprocessor/internal/common/functions.go (hunk 1 above)
Comment (rnishtala-sumo) [reply processor/transformprocessor/internal/common/functions.go:38]: Added a comment to the other PR https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23190/files#r1221861948. Will remove this

**Comment by rnishtala-sumo** [comment]
@TylerHelmuth  changing the replacement string to `StringGetter` for the `replace_all*` functions may need further changes to [ottl.Exprfunc](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c8a84c019765d25b0f70495063103f6ebd35970e/pkg/ottl/expression.go#LL18C12-L18C12)  to support a second generic type. Otherwise we run into an error like below
```
cannot use tt.replacement (variable of type ottl.StringGetter[pcommon.Value]) as type ottl.StringGetter[pcommon.Map] in argument to replaceAllPatterns[pcommon.Map]
```
I believe this is because the `replaceAllPatterns` function currently expects a `StringGetter` of type `pcommon.Map` type but the type of replacement is `StringGetter` of `pcommon.Value`.

File: processor/transformprocessor/internal/common/functions.go (hunk 1 above)
Comment (rnishtala-sumo) [reply processor/transformprocessor/internal/common/functions.go]: Removed this change

**Comment by evan-bradley** [comment]
@rnishtala-sumo If you change the type of `tt.replacement` to `StringGetter[pcommon.Map]` it should work. The type parameter in `StringGetter` just has to match the one in `replaceAllPatterns`.

**Comment by TylerHelmuth** [comment]
@rnishtala-sumo should be able to change these to StringGetter and update the tests accordingly.

- https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/d0fe5f310ca8be484e5c83c9c772144d322e5859/pkg/ottl/ottlfuncs/func_replace_all_patterns.go#L25
- https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/d0fe5f310ca8be484e5c83c9c772144d322e5859/pkg/ottl/ottlfuncs/func_replace_all_matches.go#L19

**Comment by rnishtala-sumo** [comment]
@evan-bradley @TylerHelmuth  thanks, I made the change, got a little lost in go generics there :) I tested that all of these are not breaking changes.

File: pkg/ottl/ottlfuncs/func_replace_all_matches.go [hunk 2]
```diff
@@ -33,19 +33,23 @@ func createReplaceAllMatchesFunction[K any](_ ottl.FunctionContext, oArgs ottl.A
...
 	}
 	return func(ctx context.Context, tCtx K) (interface{}, error) {
 		val, err := target.Get(ctx, tCtx)
+		replacementVal, errRepl := replacement.Get(ctx, tCtx)
 		if err != nil {
 			return nil, err
 		}
+		if errRepl != nil {
+			return nil, errRepl
+		}
```
Comment (evan-bradley) [inline pkg/ottl/ottlfuncs/func_replace_all_matches.go:49]:
```suggestion
		val, err := target.Get(ctx, tCtx)
		if err != nil {
			return nil, err
		}

		replacementVal, err := replacement.Get(ctx, tCtx)
		if err != nil {
			return nil, err
		}
```
Reassigning to `err` is more conventional, and doing the check right after the call prevents us from doing any unnecessary executions after we get an error. Not really a big deal, though, this is just a recommendation.

File: .chloggen/ottl_function_changes.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: pkg/ottl
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Changed replacement_pattern (replacement) string to be a path expression to a string telemetry field or a literal string
```
Comment (evan-bradley) [inline .chloggen/ottl_function_changes.yaml:12]: I think saying something along the lines of "change replacement functions to accept a path expression as a replacement" and listing the affected functions in the `subtext` key would be good here.

Comment (rnishtala-sumo) [reply .chloggen/ottl_function_changes.yaml]: sounds good!, made this change

**Review by TylerHelmuth — APPROVED** [review]

**Review by evan-bradley — APPROVED** [review]

**Pull request merged**